PHPCompatibility/Sniffs/Syntax/NewConstantDereferencingSniff.php

```diff
@@ -0,0 +1,135 @@
+ *
+ * As of PHP 5.6, constants can now be dereferenced to access individual elements and characters.
+ *
+ * As of PHP 8.0, constants can also be dereferenced to invoke method calls, though this support only applies to syntax and must be used in conjunction with the scalar objects extension to work at runtime.
+ *
+ * PHP version 5.6
+ * PHP version 8.0
+ *
+ * @link https://wiki.php.net/rfc/const_scalar_exprs
+ * @link https://github.com/php/php-src/commit/d5ddd2dbb263cd626a5e0f203c00d6f39168507b
+ * @link https://wiki.php.net/rfc/variable_syntax_tweaks#constant_dereferencability
+ *
+ * @since 10.0.0
+ */
+class NewConstantDereferencingSniff extends Sniff
+{
+
+    /**
+     * Returns an array of tokens this test wants to listen for.
+     *
+     * @since 10.0.0
+     *
+     * @return array
+     */
+    public function register()
+    {
+        return [
+            T_STRING,
+        ];
+    }
+
+    /**
+     * Processes this test, when one of its tokens is encountered.
+     *
+     * @since 10.0.0
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile The file being scanned.
+     * @param int                         $stackPtr  The position of the current token in
+     *                                               the stack passed in $tokens.
+     *
+     * @return void
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        // Impacted features are supported in PHP 8, bail early
+        if ($this->supportsBelow('7.4') === false) {
+            return;
+        }
+
+        $tokens = $phpcsFile->getTokens();
+
+        // Not a reference, out of scope
+        $referenceTokens  = [
+            T_OPEN_SQUARE_BRACKET => true,
+            T_OPEN_CURLY_BRACKET  => true,
+            T_OBJECT_OPERATOR     => true,
+        ];
+        $nextNonEmpty     = $phpcsFile->findNext(Tokens::$emptyTokens, ($stackPtr + 1), null, true);
+        $nextNonEmptyCode = $tokens[$nextNonEmpty]['code'];
+        if (!isset($referenceTokens[$nextNonEmptyCode])) {
+            return;
+        }
+
+        // Reference to non-constant, out of scope
+        $prevNonEmpty     = $phpcsFile->findPrevious(Tokens::$emptyTokens, ($stackPtr - 1), null, true);
+        $outOfScopeTokens = [
+            T_NAMESPACE       => true,
+            T_USE             => true,
+            T_CLASS           => true,
+            T_TRAIT           => true,
+            T_INTERFACE       => true,
+            T_EXTENDS         => true,
+            T_IMPLEMENTS      => true,
+            T_NEW             => true,
+            T_FUNCTION        => true,
+            T_OBJECT_OPERATOR => true,
+            T_INSTANCEOF      => true,
+            T_INSTEADOF       => true,
+            T_GOTO            => true,
+            T_AS              => true,
+            T_PUBLIC          => true,
+            T_PROTECTED       => true,
+            T_PRIVATE         => true,
+            'PHPCS_T_COMMA'   => true,
```

Comment:
```suggestion
            T_COMMA   => true,
```

P.S.: this should probably be removed anyhow.